File: packages/recaptcha-verification/android/src/main/kotlin/zone/picnic/recaptcha_verification/RecaptchaArgs.kt
package zone.picnic.recaptcha_verification

import io.flutter.plugin.common.MethodCall

sealed class RecaptchaArgs {
    companion object {
        fun fromCall(call: MethodCall): RecaptchaArgs {
            when (call.argument<String>("type")) {
                "webview" -> {
                    val url = call.argument<String>("url")
                        ?: throw IllegalStateException("No url")
                    val targetUrlFragment = call.argument<String>("targetUriFragment")
                        ?: throw IllegalStateException("No target uri fragment")
                    val width = call.argument<String>("width")?.toIntOrNull()
                    val height = call.argument<String>("height")?.toIntOrNull()

                    return WebView(url, targetUrlFragment, width, height)
                }
                "android" -> {
                    val siteKey = call.argument<String>("siteKey")
                        ?: throw IllegalStateException("No site key")

                    return SafetyNet(siteKey)
                }
                else -> throw IllegalStateException("Unknown captcha type")
            }
        }
    }

    data class SafetyNet(val siteKey: String) : RecaptchaArgs()
    data class WebView(
        val url: String,
        val targetUriFragment: String,
        val width: Int?,
        val height: Int?
    ) : RecaptchaArgs()
}

File: packages/recaptcha-verification/lib/src/recaptcha_args.dart
import 'package:recaptcha_verification/src/recaptcha_type.dart';

class RecaptchaArgs {
  // reCAPTCHA v2 - reCAPTCHA Android
  const RecaptchaArgs.android({required this.siteKey})
      : url = '',
        type = RecaptchaType.android,
        _targetUriFragment = '',
        width = null,
        height = null;

  // reCAPTCHA v2 - Invisible reCAPTCHA badge
  const RecaptchaArgs.iOS({required this.siteKey})
      : url = '',
        type = RecaptchaType.ios,
        _targetUriFragment = '',
        width = null,
        height = null;

  // reCAPTCHA v2 - "I'm not a robot" Checkbox
  const RecaptchaArgs.webview({
    required this.siteKey,
    this.url = '',
    this.width,
    this.height,
  })  : type = RecaptchaType.webview,
        _targetUriFragment = 'response';

  final String siteKey;
  final RecaptchaType type;
  final String url;
  final String _targetUriFragment;
  final int? width;
  final int? height;

  Map<String, String> toJson() => {
        'siteKey': siteKey,
        'url': url,
        'type': type.type,
        'targetUriFragment': _targetUriFragment,
        if (width != null) 'width': width.toString(),
        if (height != null) 'height': height.toString(),
      };
}
